### PR TITLE
feat: supported content parts

### LINKED ISSUE #156: Support content parts in chat/completion messages
Blocked by https://github.com/epam/ai-dial-sdk/issues/163

### PR DESCRIPTION
**Pull request by adubovik**
Resolves https://github.com/epam/ai-dial-adapter-bedrock/issues/156

### CODE REVIEW

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-bedrock-pr-155.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-bedrock-pr-155.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by adubovik** [comment]
/deploy-review
>Environment [URL](https://chat-ai-dial-adapter-bedrock-pr-155.nightly-test.deltixhub.io)
>E2E tests status: success

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-bedrock-pr-155.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-bedrock-pr-155.nightly-test.deltixhub.io
>E2E tests status: failed

File: aidial_adapter_bedrock/llm/message.py [hunk 1]
```diff
@@ -1,95 +1,260 @@
...
+        content = message.content
+
+        if content is None:
+            raise ValidationError("System message is expected to have content")
+
+        if isinstance(content, str):
+            return cls(content=content)
+
+        if is_text_content_parts(content):
+            return cls(content=content)  # type: ignore
```
Comment (roman-romanov-o) [inline aidial_adapter_bedrock/llm/message.py]: If we add this typeguard function here it would be intuitive to specify that we don't allow anything except text content part in system message
```python
class SystemMessage(...):
    content: str | List[MessageContentTextPart]
```

File: aidial_adapter_bedrock/llm/model/claude/v3/converters.py [hunk 2]
```diff
@@ -32,87 +35,92 @@
...
 )
 from aidial_adapter_bedrock.llm.tools.tools_config import ToolsMode
+from aidial_adapter_bedrock.utils.resource import Resource
 
 ClaudeFinishReason = Literal[
     "end_turn", "max_tokens", "stop_sequence", "tool_use"
 ]
 ImageMediaType = Literal["image/png", "image/jpeg", "image/gif", "image/webp"]
-IMAGE_MEDIA_TYPES: Set[ImageMediaType] = {
+IMAGE_MEDIA_TYPES: List[str] = [
```
Comment (roman-romanov-o) [inline aidial_adapter_bedrock/llm/model/claude/v3/converters.py:44]: Why did you move to `List[str]` from `Set[ImageMediaType]`?

File: aidial_adapter_bedrock/llm/model/llama/v3.py [hunk 3]
```diff
@@ -40,7 +40,7 @@ def encode_header(message: BaseMessage) -> str:
 
 def encode_message(message: BaseMessage) -> str:
     ret = encode_header(message)
-    ret += message.content.strip()
+    ret += message.text_content
```
Comment (roman-romanov-o) [inline aidial_adapter_bedrock/llm/model/llama/v3.py]: I guess, if we used strip before, let's be consistent
```python
message.text_content.strip()
```

File: aidial_adapter_bedrock/dial_api/request.py [hunk 4]
```diff
@@ -51,3 +56,39 @@ def tools_mode(self) -> ToolsMode | None:
         if self.tool_config is not None:
             return self.tool_config.tools_mode
         return None
+
+
+def collect_text_content(
+    content: str | List[MessageContentPart] | None,
+    delimiter: str = "\n",
```
Comment (roman-romanov-o) [inline aidial_adapter_bedrock/dial_api/request.py]: I saw some examples, where different parts of content were separated by more verbose `\n\n`

Maybe we should do the same?

p.s.: I don't have some test result to prove that's affecting model response though

File: aidial_adapter_bedrock/dial_api/resource.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+    def create_url_resource(self, url: str) -> URLResource:
+        return URLResource(
+            url=url,
+            content_type=self.effective_content_type,
+            entity_name=self.entity_name,
+        )
+
+    @property
+    def effective_content_type(self) -> str | None:
```
Comment (roman-romanov-o) [inline aidial_adapter_bedrock/dial_api/resource.py]: What do you mean by `effective`?

File: aidial_adapter_bedrock/dial_api/resource.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+
+class AttachmentResource(DialResource):
+    attachment: Attachment
+
+    @validator("attachment", pre=True)
+    def parse_attachment(cls, value):
+        if isinstance(value, dict):
+            attachment = Attachment.parse_obj(value)
+            # Working around the issue of defaulting missing type to a markdown:
```
Comment (roman-romanov-o) [inline aidial_adapter_bedrock/dial_api/resource.py:112]: Could you elaborate a bit more on this?

Because, if I understand correctly, defaulting to markdown should be in this part
```python
    async def guess_content_type(self) -> str | None:
        return (
            self.content_type
            or Resource.parse_data_url_content_type(self.url)
            or mimetypes.guess_type(self.url)[0] <----
        )
```
right?

**Review by roman-romanov-o — COMMENTED** [review]
.

File: aidial_adapter_bedrock/llm/message.py (hunk 1 above)
Comment (adubovik) [reply aidial_adapter_bedrock/llm/message.py]: Fixed

File: aidial_adapter_bedrock/llm/model/claude/v3/converters.py (hunk 2 above)
Comment (adubovik) [reply aidial_adapter_bedrock/llm/model/claude/v3/converters.py:44]: The order of keys in the set isn't fixed from one program run to another. Since the set is printed in the user-facing error message, I need to remember to call `sorted` before printing the list of image types, so as to make sure that the error message is always the same. 

With a list I don't need to do it.

Besides, the lookup performance in short lists is comparable to the one in sets.

File: aidial_adapter_bedrock/llm/model/llama/v3.py (hunk 3 above)
Comment (adubovik) [reply aidial_adapter_bedrock/llm/model/llama/v3.py]: Indeed, fixed

File: aidial_adapter_bedrock/dial_api/request.py (hunk 4 above)
Comment (adubovik) [reply aidial_adapter_bedrock/dial_api/request.py]: Fixed

File: aidial_adapter_bedrock/dial_api/resource.py (hunk 5 above)
Comment (adubovik) [reply aidial_adapter_bedrock/dial_api/resource.py]: Effective means informative. Catch-all binary content type is not informative, so it's discounted in this method.

File: aidial_adapter_bedrock/dial_api/resource.py (hunk 6 above)
Comment (adubovik) [reply aidial_adapter_bedrock/dial_api/resource.py:112]: It's about working-around an unfortunate oversight in DIAL SDK itself:

https://github.com/epam/ai-dial-sdk/blob/2835107e950c89645a2b619fecba2518fa2d7bb1/aidial_sdk/chat_completion/request.py#L22

it defaults a missing `type` field to `text/markdown`, which is surprising:

```python
Attachment.parse_obj({"url": "doc.pdf"}).type == "text/markdown"
```

Comment (adubovik) [reply aidial_adapter_bedrock/dial_api/resource.py:112]: I left the link to DIAL SDK in the comment.

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-bedrock-pr-155.nightly-test.deltixhub.io
>E2E tests status: success

File: aidial_adapter_bedrock/dial_api/resource.py (hunk 5 above)
Comment (roman-romanov-o) [reply aidial_adapter_bedrock/dial_api/resource.py]: Maybe let's rename it to `informative_content_type` then?

Or at least leave some comment about this

Because `effective` adjective in terms of content type may mislead you in thinking about encoding effectiveness itself

Comment (roman-romanov-o) [reply aidial_adapter_bedrock/dial_api/resource.py]: Or we can use `meaningful_content_type` for that, what do you think?

Comment (adubovik) [reply aidial_adapter_bedrock/dial_api/resource.py]: Renamed to `informative_content_type`

**Review by roman-romanov-o — APPROVED** [review]
